mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,176 @@
+        Number of Legendre terms to evaluate.
+    sphere : array-like, shape (4,)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius in meters.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if inst.info['custom_ref_applied'] == -1:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    inst = inst.copy() if copy else inst
+
+    picks = pick_types(inst.info, meg=False, eeg=True, exclude='bads')
+
+    if len(picks) == 0:
+        raise ValueError('No EEG channels found.')
+
+    if lambda2 is None:
+        lambda2 = 1e-5
+
+    _validate_type(lambda2, 'numeric', 'lambda2')
+    if 0 > lambda2 or lambda2 > 1:
+        raise ValueError('lambda2 must be between 0 and 1, got %s' % lambda2)
+
+    _validate_type(stiffness, 'numeric', 'stiffness')
+    if stiffness < 0:
+        raise ValueError('stiffness must be non-negative got %s' % stiffness)
+
+    n_legendre_terms = _ensure_int(n_legendre_terms, 'n_legendre_terms')
+    if n_legendre_terms < 1:
+        raise ValueError('n_legendre_terms must be greater than 0, '
+                         'got %s' % n_legendre_terms)
+
+    if sphere == 'auto':
+        radius, origin_head, origin_device = fit_sphere_to_headshape(inst.info)
+        x, y, z = origin_head - origin_device
+    else:
+        _validate_type(sphere, tuple, 'sphere')
+        x, y, z, radius = sphere
+
+    _validate_type(x, 'numeric', 'x')
+    _validate_type(y, 'numeric', 'y')
+    _validate_type(z, 'numeric', 'z')
+    _validate_type(radius, 'numeric', 'radius')
+    if radius <= 0:
+        raise ValueError('sphere radius must be greater than 0, '
+                         'got %s' % radius)
+
+    _validate_type(copy, (bool), 'copy')
+
+    if not _check_ch_locs(inst.info['chs']):
+        raise ValueError('Zero or infinite position found in chs')
+
+    pos = np.array([inst.info['chs'][pick]['loc'][:3] for pick in picks])
+    pos -= (x, y, z)
+
+    G = _calc_g(np.dot(pos, pos.T), stiffness=stiffness,
+                num_lterms=n_legendre_terms)
+    H = _calc_h(np.dot(pos, pos.T), stiffness=stiffness,
+                num_lterms=n_legendre_terms)
+
+    G_precomputed = _prepare_G(G, lambda2)
+
+    trans_csd = _compute_csd(G_precomputed=G_precomputed,
+                             H=H, radius=radius)
+
+    if isinstance(inst, BaseEpochs):
+        for epo in inst._data:
+            epo[picks] = np.dot(trans_csd, epo[picks])
+    else:
+        inst._data = np.dot(trans_csd, inst._data[picks])
+
+    pick_info(inst.info, picks, copy=False)
+    inst.info['custom_ref_applied'] = -1
+    for ch in inst.info['chs']:
+        if ch['coil_type'] == FIFF.FIFFV_COIL_EEG:
```

Comment:
Eventually, once we process good and bad channels, you can just iterate over `for pick in picks: info['chs'][pick] ...`